src/ray/util/event.cc

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/event.h"
+
+namespace ray {
+thread_local std::unique_ptr<RayEventContext> RayEventContext::instance_ = nullptr;
+
+LogBasedEventReporter::LogBasedEventReporter(rpc::Event_SourceType source_type, std::string log_dir)
+    : log_dir_(log_dir) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+}
+
+std::string LogBasedEventReporter::EventToString(rpc::Event &event) {
+  std::stringstream result;
+
+  struct tm *dt;
+  char time_buffer[30];
+  time_t epoch_time_as_time_t = event.timestamp() / 1000;
+
+  dt = localtime(&epoch_time_as_time_t);
+  strftime(time_buffer, sizeof(time_buffer), "%Y-%m-%d %H:%M:%S.", dt);
+
+  result << std::string(time_buffer) << std::setw(5) << std::setfill('0')
+         << event.timestamp() % 1000 << separator_ << event.event_id() + separator_
+         << event.job_id() + separator_ << event.node_id() + separator_
+         << event.task_id() + separator_
+         << Event_SourceType_Name(event.source_type()) + separator_
+         << event.source_hostname() + separator_
+         << std::to_string(event.source_pid()) + separator_
+         << Event_Severity_Name(event.severity()) + separator_
```

Comment:
Why don't we make it like json format, so that it is easy to parse?

Reply:
First, this function is a virtual function that can be inherited. BTW, If we intend to write files, The JSON format will write more bytes than the current implementation. In our internal system, the monitor backend prefers to read the separator to analyze the event file.

Reply:
We should just use JSON as a standard format instead of inventing our own. This will make it easier for OSS users to integrate with and is similar to other structured logging formats (e.g., https://github.com/uber-go/zap)

Reply:
We should not use the format that is suitable for Ant's internal systems in OSS. It should be the format that is more generally used by the industry. (I guess you guys can have a separate implementation to use this separator).

Reply:
@edoakes I think the example of zap is convinced for us. We plan to change to the JSON format.
I think there is a small question about If the user prints multiple lines for example:
RAY_EVENT(INFO, "label") << "process 1\nprocess 2\nprocess3"
How do we solve this situation, should we print multiple lines?